presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java

```diff
@@ -1371,6 +1372,11 @@ public SystemSessionProperties(
                         "Use perfectly consistent histories for history based optimizations, even when parts of a query are re-ordered.",
                         featuresConfig.isUsePerfectlyConsistentHistories(),
                         false),
+                integerProperty(
+                        HISTORIES_LIMITING_PLAN_NODE_LIMIT,
+                        "Use history based optimizations only when cost of hashing \"limiting\" plan nodes like TopN, Limit, DistinctLimit is within this limit. This limit will be directly proportional to the maximum number of plan nodes optimizer can handle.",
```

Comment:
this isn't super clear for users to be able to understand what to set here. trying to think of a better name/description.

Reply:
What do you think about history_canonical_plan_node_limit? Since it will be the "effective" size of canonical plan. We can remove "limiting" term here, and just say for some cases it may become large, and if its too large - we skip hbo

Reply:
Let me know what you think about this

Reply:
I believe this can also be just a config(not a session prop) - but want it like this for now so its easy to experiment and tune it